### PR TITLE
[ENH] Added useful class attributes to SAST based transforms

### LINKED ISSUE #2004: [ENH] Shapelet transforms vary in class attributes
### Describe the feature or idea you want to propose

Overarching idea is to make the shapelet transforms implemented more similarly to one another, but for now they should be made to have  the same class attributes for example shapelet transform has useful class attributes such as (shapelet information gain, shapelet length, start position the shapelet was extracted from, shapelet dimension, index of the instance the shapelet was extracted from in fit, class value of the shapelet. These are all very helpful for shapelet interpretation but aren't always present in the other transforms.

### Describe your proposed solution

Add missing class attributes to SAST,RSAST (RDST has already been done I think). Though these attributes may not serve any purpose in the functionality of the algorithms, hence why they probably weren't added to the class, they are useful for comparing the shapelets generated by each transform,

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by IRKnyazev**
#### Reference Issues/PRs
Fixes #2004 

#### What does this implement/fix? Explain your changes.
Added class attributes found in ST that are helpful for interpreting shapelets. They do not serve functionality for the classifier but improve explainability.

Implemented:
The class each subseries is from, 
the index of the original series the subseries is extracted from, 
the start pos of the subseries in the original series.

### CODE REVIEW

File: aeon/transformations/collection/shapelet_based/_rsast.py [hunk 1]
```diff
@@ -63,11 +64,28 @@ class RSAST(BaseCollectionTransformer):
         the number of reference time series to select per class
     seed : int, default = None
         the seed of the random generator
-    classifier : sklearn compatible classifier, default = None
+    estimator : sklearn compatible classifier, default = None
         if None, a RidgeClassifierCV(alphas=np.logspace(-3, 3, 10)) is used.
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/shapelet_based/_rsast.py:68]: This does not exist in the transformer by the look of things?

File: aeon/classification/shapelet_based/_sast.py [hunk 2]
```diff
@@ -178,6 +178,7 @@ def plot_most_important_feature_on_ts(self, ts, feature_importance, limit: int =
         """
         import matplotlib.pyplot as plt
 
+        feature_importance = abs(feature_importance)
```
Comment (MatthewMiddlehurst) [inline aeon/classification/shapelet_based/_sast.py:181]: Does this make sense to do? I'm not familiar with what metric is used here.

File: aeon/transformations/collection/shapelet_based/_rsast.py [hunk 3]
```diff
@@ -63,11 +64,28 @@ class RSAST(BaseCollectionTransformer):
...
+    _start_positions : list
+        The starting positions of each subsequence within the original time series.
+    _classes : list
+        The class labels associated with each subsequence.
+    _source_series: list
+        The index of the original time series in the training set from which each
+        subsequence was derived.
+    _kernels_generators_ : dict
+        A dictionary mapping class labels to the selected reference time series
+        for that class.
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/shapelet_based/_rsast.py:87]: Attributes starting with _ are not public and dont need to be included here. Ending with _ means set in fit and public.

Comment (IRKnyazev) [reply aeon/transformations/collection/shapelet_based/_rsast.py:87]: ohh I see, thanks Matthew!

File: aeon/classification/shapelet_based/_sast.py (hunk 2 above)
Comment (IRKnyazev) [reply aeon/classification/shapelet_based/_sast.py:181]: There are two mentions to making feature importance abs in the SASt paper, this is specifically needed for linear classifiers. 
image
image
I could contact Michael to confirm if you would like?

File: aeon/transformations/collection/shapelet_based/_rsast.py (hunk 1 above)
Comment (IRKnyazev) [reply aeon/transformations/collection/shapelet_based/_rsast.py:68]: True, the docstring must have been copied across from the RSAST classifier originally. Will remove this whole bit

File: aeon/classification/shapelet_based/_sast.py (hunk 2 above)
Comment (IRKnyazev) [reply aeon/classification/shapelet_based/_sast.py:181]: @baraline do you think we can remove the plot_most_important_feature_on_ts. It seems it was implemented before your shapelet viz module, and this functionality can be achieved there now instead?

Comment (baraline) [reply aeon/classification/shapelet_based/_sast.py:181]: It should be. Although a quick test using both modules won't hurt to ensure that they both produce the same kind of visualization.

**Comment by baraline** [comment]
Hi Ivan, while memory wise the storage of these parameters would likely not cause too much trouble, I feel like these kind of optional storage should be gated behind a boolean given during initialization (something like `store_shapelet_intialization_info` ?).

I don't know if that's the goal of the PR (if not ignore it), but preferably, all the different shapelets methods should have the same name for these arguments, so we can use a common interface while inspecting them.

**Comment by IRKnyazev** [comment]
> Hi Ivan, while memory wise the storage of these parameters would likely not cause too much trouble, I feel like this kind of optional storage should be gated behind a boolean given during initialization (something like `store_shapelet_intialization_info` ?).
I don't think having these extra private variables will cause confusion for users, I  think it is more straight forward having them there for accessibility rather than providing the option for storing them. More than happy to implement your idea though if you disagree. Maybe @MatthewMiddlehurst might have a preference?


> I don't know if that's the goal of the PR (if not ignore it), but preferably, all the different shapelets methods should have the same name for these arguments, so we can use a common interface while inspecting them.

Great point, I modified the start point variable name to be consistent with the other transforms!

File: aeon/transformations/collection/shapelet_based/_rsast.py [hunk 4]
```diff
@@ -187,11 +194,15 @@ def _fit(self, X: np.ndarray, y: Union[np.ndarray, list]) -> "RSAST":
...
+            # Store the original indices of the sampled time series
+            original_indices = np.where(y == c)[0]
+
+            chosen_indices = self._random_state.permutation(X_c.shape[0])[:cnt]
 
             self._kernels_generators[c] = []
 
-            for rep, idx in enumerate(choosen):
+            for rep, idx in enumerate(chosen_indices):
+                original_idx = original_indices[idx]  # Get the original index
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/shapelet_based/_rsast.py:205]: dont really need this line, can just add `original_indices[idx]`

**Review by MatthewMiddlehurst — COMMENTED** [review]
Given it's private I don't think theres a harm in adding it assuming the memory footprint is not large.

I agree we should eventually unify this between the approaches if possible. Not sure how they all do it differently, I think STC is a list of lists for each shapelet.

Happy to remove the function if it is mostly duplicated by the new visualisation tools.

**Review by TonyBagnall — APPROVED** [review]

**Pull request merged**